evm/src/memory/memory_stark.rs

```diff
@@ -321,17 +334,23 @@ impl<F: RichField + Extendable<D>, const D: usize> MemoryStark<F, D> {
         trace_cols[VIRTUAL_FIRST_CHANGE] = virtual_first_change;
 
         trace_cols[RANGE_CHECK] = range_check_value;
-        trace_cols[COUNTER] = (0..num_trace_rows)
-            .map(|i| F::from_canonical_usize(i))
-            .collect();
+
+        for col in COLUMNS_TO_PAD {
+            trace_cols[col].splice(0..0, vec![F::ZERO; to_pad]);
+        }
+
+        trace_cols[COUNTER] = (0..to_pad_to).map(|i| F::from_canonical_usize(i)).collect();
 
         let (permuted_inputs, permuted_table) =
             permuted_cols(&trace_cols[RANGE_CHECK], &trace_cols[COUNTER]);
         trace_cols[RANGE_CHECK_PERMUTED] = permuted_inputs;
         trace_cols[COUNTER_PERMUTED] = permuted_table;
     }
 
-    pub fn generate_trace(&self, memory_ops: Vec<MemoryOp<F>>) -> Vec<PolynomialValues<F>> {
+    pub fn generate_trace(
+        &self,
+        memory_ops: Vec<MemoryOp<F>>,
+    ) -> (Vec<PolynomialValues<F>>, usize) {
```

Comment:
Could the caller get the number of ops based on the length of the first polynomial instead? Would prefer to avoid changing this method signature if we can, since it matches similar methods for the other tables.

Reply:
Oh yes I was wondering about that, good point